### PR TITLE
Todo MVC 

### PR DESCRIPTION
**Pull request by vramana**

### CODE REVIEW

**Comment by tsingson** [comment]
great, the example should full for us to try redux.

File: examples/components/AddTodo.js [hunk 1]
```diff
@@ -5,15 +5,19 @@ export default class AddTodo {
     addTodo: PropTypes.func.isRequired
   };
 
+  inputHandler(e) {
+    const text = this.refs.input.getDOMNode().value;
+    if (e.which === 13 && text.length !== 0) {
+      this.props.addTodo(text);
+      this.refs.input.getDOMNode().value = '';
```
Comment (ooflorent) [inline examples/components/AddTodo.js]: `getDOMNode` is deprecated. You must use `React.findDOMNode`.

Comment (vramana) [inline examples/components/AddTodo.js]: @ooflorent I didn't know that.

File: examples/actions/TodoActions.js [hunk 2]
```diff
@@ -6,3 +6,17 @@ export function addTodo(text) {
     text
   };
 }
+
+export function delTodo(id) {
```
Comment (ooflorent) [inline examples/actions/TodoActions.js]: I don't like this naming. `deleteTodo` would be more expressive.

**Comment by vramana** [comment]
@ooflorent I have restructed the components. Can you take a look at this ??

File: examples/components/TodoItem.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                onChange={()=> markTodo(todo.id)}
+            />
+            <label onDoubleClick={() => this.onDoubleClick()}>{todo.text}</label>
+            <button
+              className='destroy'
+              onClick={()=> deleteTodo(todo.id)}
+            />
+        </div>
+        <TodoTextInput
+            class='edit'
```
Comment (gaearon) [inline examples/components/TodoItem.js]: What does `class='edit'` mean? Why is this a string?

File: examples/components/TodoTextInput.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      node.focus();
+      node.setSelectionRange(node.value.length, node.value.length);
+    }
+  }
+
+  submit(e) {
+    const text = e.target.value.trim();
+    if (e.which === 13) {
+      this.props.onSave(text);
+      if (this.props.id === 'new-todo') {
```
Comment (gaearon) [inline examples/components/TodoTextInput.js]: Why not use null `id` for new todo instead of a string?

File: examples/components/TodoItem.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    }
+    this.setState({ edit: false });
+  }
+
+  render() {
+    const {todo, markTodo, deleteTodo, editTodo} = this.props;
+    return (
+      <li className={classnames({
+            completed: todo.marked,
+            editing: this.state.edit
```
Comment (ooflorent) [inline examples/components/TodoItem.js]: I'm not really fond of this.
I'd prefer if the view to render only what is needed.

File: examples/components/TodoTextInput.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    this.state = {
+      text: this.props.text || ''
+    };
+  }
+
+  componentDidUpdate() {
+    // Sets the cursor position at the end of todo.
+    // edit variable is used to keep focus on new-todo otherwise new-todo loses
+    // focus after entering a todo.
+    if (this.props.class === 'edit' && this.props.edit) {
```
Comment (ooflorent) [inline examples/components/TodoTextInput.js]: Do you mean `className`?
I think this is a bad practice. Its parent could pass an `edit` property to handle this case.

File: examples/components/TodoTextInput.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    return (
+      <input
+          className={this.props.class}
+          id={this.props.id}
+          type='text'
+          placeholder={this.props.placeholder}
+          autoFocus='true'
+          value={this.state.text}
+          onBlur={(e) => this.props.onSave(e.target.value)}
+          onChange={(e) => this.onChange(e)}
```
Comment (ooflorent) [inline examples/components/TodoTextInput.js]: What about `::this.onChange`? IMHO it is more readable.

**Comment by ooflorent** [comment]
I think each example should be contained by its own directory thus allowing a `package.json` per example not to pollute the main one.

File: examples/components/TodoTextInput.js (hunk 7 above)
Comment (vramana) [inline examples/components/TodoTextInput.js]: I am not fully familiar with bind syntax. I will change it.

**Comment by vramana** [comment]
@ooflorent I have made the changes and also moved the example to it's own folder. I temporarily removed the counter example. I will add it back.

File: examples/todomvc/components/TodoItem.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+  render() {
+    const {todo, markTodo, deleteTodo, editTodo} = this.props;
+
+    let view;
+
+    if (this.state.edit) {
+      view = (<TodoTextInput
+                  text={todo.text}
+                  edit={this.state.edit}
```
Comment (ooflorent) [inline examples/todomvc/components/TodoItem.js]: Not sure about that. `state.edit` is always `true` is this branch.

File: examples/todomvc/components/TodoTextInput.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
+import React, { Component, PropTypes } from 'react';
+import classnames from 'classnames';
+
+export default class TodoTextInput extends Component {
+  static propTypes = {
+    onSave: PropTypes.func.isRequired
```
Comment (ooflorent) [inline examples/todomvc/components/TodoTextInput.js]: Missing some props (`text`, `onSave`, etc)

**Comment by ooflorent** [comment]
It looks promising :)

File: examples/todomvc/components/TodoItem.js (hunk 8 above)
Comment (vramana) [inline examples/todomvc/components/TodoItem.js]: `this.state.edit` is only true on double click. When the `onsave` function triggers it sets the `edit` to `false` again

File: examples/todomvc/components/TodoItem.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    todo: PropTypes.object.isRequired,
+    editTodo: PropTypes.func.isRequired,
+    deleteTodo: PropTypes.func.isRequired,
+    markTodo: PropTypes.func.isRequired
+  };
+
+  constructor(props, context) {
+    super(props, context);
+    this.state = {
+      edit: false
```
Comment (gaearon) [inline examples/todomvc/components/TodoItem.js]: Rename that to `editing` maybe?

File: examples/todomvc/components/TodoItem.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    this.setState({ edit: false });
+  }
+
+  render() {
+    const {todo, markTodo, deleteTodo, editTodo} = this.props;
+
+    let view;
+
+    if (this.state.edit) {
+      view = (<TodoTextInput
```
Comment (gaearon) [inline examples/todomvc/components/TodoItem.js]: Style nit pick: please put newline after the `(` in JSX:

```js
  view = (
    <TodoTextInput attr=...
                   anotherAttr=... />
  );
```

This way they're all intended the same.

Also please change `view` to `element` or `input`.. Don't want to sound MVCish there.

File: examples/todomvc/components/TodoItem.js [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+                  text={todo.text}
+                  edit={this.state.edit}
+                  onSave={(text) => this.onSave(todo.id, text)} />);
+    } else {
+      view = (<div className='view'>
+                <input
+                    className='toggle'
```
Comment (gaearon) [inline examples/todomvc/components/TodoItem.js]: Nit pick: here and after please put the first attribute on the same line.

```js
<input className='toggle'
       type='...'
       lala='..' />
```

File: examples/todomvc/stores/todos.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+      return state.map(todo => {
+        if (todo.id === action.id) {
+          todo.marked = !todo.marked;
+        }
+        return todo;
+      });
+    case MARK_ALL: {
+      if (state.filter(todo => todo.marked).length < state.length) {
+        return state.map(todo => {
+          todo.marked = true;
```
Comment (gaearon) [inline examples/todomvc/stores/todos.js]: This is mutation. Please avoid mutations. Write `todo => { ...todo, marked: true }` instead.

File: examples/todomvc/components/TodoItem.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+                <button
+                  className='destroy'
+                  onClick={() => deleteTodo(todo.id)}
+                />
+              </div>);
+    }
+
+    return (
+      <li className={classnames({
+            completed: todo.marked,
```
Comment (gaearon) [inline examples/todomvc/components/TodoItem.js]: Nitpick: please align these to be two spaces from `<`:

```js
return (
  <li lala={lala({
    lala: true
  })}>
    ...
  </li>
)
)

File: examples/todomvc/containers/App.js [hunk 15]
```diff
@@ -11,12 +10,7 @@ export default class App {
   render() {
     return (
       <Provider redux={redux}>
-        {() =>
-          <div>
-            <CounterApp />
-            <TodoApp />
-          </div>
-        }
+        { () => <TodoApp /> }
```
Comment (gaearon) [inline examples/todomvc/containers/App.js]: Nit: no need for extra spaces here. `{() => <TodoApp />}`

File: examples/todomvc/index.js [hunk 16]
```diff
@@ -1,5 +1,6 @@
 import React from 'react';
 import App from './containers/App';
+import 'style!raw!./node_modules/todomvc-app-css/index.css';
```
Comment (gaearon) [inline examples/todomvc/index.js]: Won't just `style!raw!todomvc-app-css/index.css` work? This `node_modules` thing looks weird.

File: examples/todomvc/stores/todos.js (hunk 13 above)
Comment (vramana) [inline examples/todomvc/stores/todos.js]: Does any explicit written `(key, value1)` pair replace the destructured `(key, value2)`  regardless of the order ?

Comment (gaearon) [inline examples/todomvc/stores/todos.js]: Not sure I understand the question but this should help: https://github.com/sebmarkbage/ecmascript-rest-spread/blob/master/Spread.md

**Comment by vramana** [comment]
@gaearon Done. Anymore nitpicks?

File: examples/todomvc/components/Header.js [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+import TodoTextInput from './TodoTextInput';
+
+export default class Header {
+
+  render() {
+    return (
+      <header className='header'>
+          <h1>todos</h1>
+          <TodoTextInput newTodo={true}
+                         onSave={(text) => (text.length !== 0) ? this.props.addTodo(text) : undefined}
```
Comment (ooflorent) [inline examples/todomvc/components/Header.js]: Don't use `undefined`

File: examples/todomvc/components/TodoTextInput.js [hunk 18]
```diff
@@ -0,0 +1,46 @@
+import React, { Component, PropTypes } from 'react';
+import classnames from 'classnames';
+
+export default class TodoTextInput extends Component {
+  static propTypes = {
+    onSave: PropTypes.func.isRequired,
```
Comment (ooflorent) [inline examples/todomvc/components/TodoTextInput.js]: Missing `editing`, `placeholder`, `newTodo`

File: examples/todomvc/components/TodoTextInput.js [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+  };
+
+  constructor(props, context) {
+    super(props, context);
+    this.state = {
+      text: this.props.text || ''
+    };
+  }
+
+  submit(e) {
```
Comment (ooflorent) [inline examples/todomvc/components/TodoTextInput.js]: Can you unify event listener names? `submit` and `onChange` are both listeners but don't follow the same naming convention

File: examples/todomvc/components/TodoItem.js [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+    this.setState({ editing: false });
+  }
+
+  render() {
+    const {todo, markTodo, deleteTodo, editTodo} = this.props;
+
+    let element;
+
+    if (this.state.editing) {
+
```
Comment (ooflorent) [inline examples/todomvc/components/TodoItem.js]: Spacing is weird

File: examples/todomvc/components/Header.js (hunk 17 above)
Comment (gaearon) [inline examples/todomvc/components/Header.js]: I'd rather have `handleSave` as a class method here.

File: examples/todomvc/components/TodoTextInput.js (hunk 19 above)
Comment (gaearon) [inline examples/todomvc/components/TodoTextInput.js]: Class methods that are event handlers need to have `handle*` names.

**Comment by vramana** [comment]
@ooflorent I don't know why the tests are failing on travis. But all the tests pass.

File: examples/todomvc/stores/todos.js [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+        id: (state.length === 0) ? 0 : state[0].id + 1,
+        marked: false,
+        text: action.text
+      }, ...state];
+
+    case DELETE_TODO:
+      return state.filter(todo => todo.id !== action.id);
+
+    case EDIT_TODO:
+      return state.map(todo => todo.id === action.id ? { ...todo, text: action.text } : todo);
```
Comment (gaearon) [inline examples/todomvc/stores/todos.js]: Let's break this long line (and other similar ones)

```
return state.map(todo =>
  todo.id === action.id ?
    { ...todo, text: action.text } :
    todo
);

**Comment by vramana** [comment]
@gaearon I made the changes you have asked.

File: examples/todomvc/components/MainSection.js [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+
+  render() {
+    return (
+      <section className='main'>
+        <input
+            className='toggle-all'
+            type='checkbox'
+            onChange={::this.props.actions.markAll} />
+        <ul className='todo-list'>
+          {this.props.todos.map(todo => <TodoItem key={todo.id} todo={todo} {...this.props.actions} />)}
```
Comment (gaearon) [inline examples/todomvc/components/MainSection.js]: Can you please spit a line after `=>` ?

**Comment by gaearon** [comment]
This looks really clean and nice to me. Great job.

Can somebody else please review and merge this?
I want to make sure package.json and stuff like that is correct, but I haven't got time atm. @ooflorent @acdlite @emmenko ?

**Comment by gaearon** [comment]
Can you please squash into a single commit in the meantime?

**Comment by gaearon** [comment]
What does `npm start` in the root project do now? It used to launch the example. Should it be removed?

**Comment by vramana** [comment]
@gaearon I guess so. Let me also add the temporarily deleted counter example while at it.

**Comment by gaearon** [comment]
:+1:

**Comment by gaearon** [comment]
Please also merge master in (sorry for the churn!) as it contains the new ESLint config from #121. Please fix any style issues coming from linting with the new config.

**Comment by gaearon** [comment]
This is amazing effort. Thank you so much.

**Comment by emmenko** [comment]
@vramana wow, it looks awesome. Great work!

**Comment by vramana** [comment]
@emmenko Thanks !!

File: examples/todomvc/components/TodoTextInput.js [hunk 23]
```diff
@@ -0,0 +1,49 @@
+import React, { Component, PropTypes } from 'react';
+import classnames from 'classnames';
+
+export default class TodoTextInput extends Component {
+  static propTypes = {
+    handleSave: PropTypes.func.isRequired,
```
Comment (ooflorent) [inline examples/todomvc/components/TodoTextInput.js]: I hadn't the time to check this before it has been merged.

The props referring to events should be named `on[EventName]`.
The handlers should be named `handle[EventName]`.

Can you rename it to `onSave`?

Comment (vramana) [inline examples/todomvc/components/TodoTextInput.js]: Okay. I will do that.

Comment (gaearon) [inline examples/todomvc/components/TodoTextInput.js]: Good catch!

**Pull request merged**